lib/database/availibility-feed.ts

```diff
@@ -0,0 +1,141 @@
+ * an array of intervals representing all the times in between start and end that were not in
+ * the passed in intervals array
+ *
+ * Think of this as the "not" operator being applied to the passed in intervals (getting the oppositve of all intervals)
+ */
+const calculateBetweenIntervals = (
+  start: DateTime,
+  end: DateTime,
+  intervals: Interval[]
+): Interval[] => {
+  if (intervals.length == 0) {
+    return [Interval.fromDateTimes(start, end)];
+  }
+
+  const newIntervals = intervals.map((interval, index) => {
+    if (index == 0) return Interval.fromDateTimes(start, interval.start);
+    return Interval.fromDateTimes(intervals[index - 1].end, interval.start);
+  });
+  newIntervals.push(Interval.fromDateTimes(intervals[intervals.length - 1].end, end));
+  return newIntervals.filter(
+    (interval) => interval.length("milliseconds") != 0 && interval.isValid
+  );
+};
+
+// get all intervals with dates
+// calculate unavailibilities
+// use merge from luxon on unavaibililites and events
+// reverse unavailibilities
+
+const getAvailibilityFeed = async (
+  start: string,
+  end: string,
+  userId: string
+): Promise<CalendarEvent[]> => {
+  const dates = getDatesInInterval(start, end);
+  const user = await getUser(userId);
+  if (user == null) {
+    throw Error("Could not retrieve user from database");
+  }
+  const availibility = await getAvailibilityById(userId);
+  if (availibility == null) {
+    throw Error("Could not fetch availibility for user");
+  }
+  const availibilityAsIntervals: Interval[] = [];
+  const processedDaysOfWeek = new Set<string>();
+  dates.forEach((date) => {
+    const weekdayStr = indexToWeekdays[date.weekday] as Weekdays;
+    if (processedDaysOfWeek.has(weekdayStr) || weekdayStr == "sun") return;
+
+    processedDaysOfWeek.add(weekdayStr);
+    let availibilitiesToProcess = availibility[weekdayStr] == null ? [] : availibility[weekdayStr];
+    if (availibilitiesToProcess == null) availibilitiesToProcess = [];
+    availibilitiesToProcess.forEach((availibilityInterval) => {
+      const availibilityWithDate = insertDateIntoInterval(
+        availibilityInterval,
+        date.toJSDate(),
+        availibility.timeZone
+      );
+      availibilityAsIntervals.push(
+        Interval.fromDateTimes(availibilityWithDate[0], availibilityWithDate[1])
+      );
+    });
+  });
+
+  const unavaibililites = calculateBetweenIntervals(
+    DateTime.fromISO(start),
+    DateTime.fromISO(end),
+    availibilityAsIntervals
+  );
+  const userEvents = (await getEventFeed(start, end, userId)).map((event) => {
+    return Interval.fromDateTimes(DateTime.fromISO(event.start), DateTime.fromISO(event.end));
+  });
+
+  const completeUnavailibilites = unavaibililites.concat(userEvents);
+
+  const mergedUnavilibilites = Interval.merge(completeUnavailibilites);
+
+  const finalAvailibility = calculateBetweenIntervals(
+    DateTime.fromISO(start),
+    DateTime.fromISO(end),
+    mergedUnavilibilites
```

Comment:
Are these mergedUnavailabilities guaranteed to be sorted? If not I think that would break calculateBetweenIntervals()